frontend/lib/resolution/disambiguation.cpp

```diff
@@ -1213,8 +816,215 @@ computeIsMoreVisible(Context* context,
+    return nullptr;
+  }
+  if (candidates.size() == 1) {
+    return candidates[0];
+  }
+
+  // Disable implicit conversion to remove nilability for disambiguation
+  // int saveGenerousResolutionForErrors = 0;
+  // if (generousResolutionForErrors > 0) {
+  //   saveGenerousResolutionForErrors = generousResolutionForErrors;
+  //   generousResolutionForErrors = 0;
+  // }
+
+  // If index i is set we have ruled out that function
+  std::vector<bool> discarded(candidates.size(), false);
+
+  // use new rules
+  disambiguateDiscarding(dctx, candidates, ignoreWhere, discarded);
+
+  // If there is just 1 candidate at this point, return it
+  {
+    int only = -1;
+    int currentCandidates = 0;
+    for (size_t i = 0; i < candidates.size(); ++i) {
+      if (discarded[i]) {
+        continue;
+      }
+      only = i;
+      currentCandidates++;
+    }
+
+    if (currentCandidates == 1) {
+      EXPLAIN("Y: Fn %d is the best match.\n\n\n", only);
+
+      // if (saveGenerousResolutionForErrors > 0)
+      //   generousResolutionForErrors = saveGenerousResolutionForErrors;
+
+      return candidates[only];
+    }
+  }
+
+  // There was more than one best candidate.
+  // So, add whatever is left to 'ambiguous'
+  // and return NULL.
+  for (size_t i = 0; i < candidates.size(); i++) {
+    if (discarded[i]) {
+      continue;
+    }
+
+    EXPLAIN("Z: Fn %lu is one of the best matches\n", i);
+    ambiguous.push_back(candidates[i]);
+  }
+
+  // if (saveGenerousResolutionForErrors > 0)
+  //   generousResolutionForErrors = saveGenerousResolutionForErrors;
+
+  return nullptr;
+}
+
+static const DisambiguationCandidate*
+disambiguateByMatch(const DisambiguationContext& dctx,
+                    const CandidatesVec& candidates,
+                    bool ignoreWhere,
+                    CandidatesVec& ambiguousBest) {
+
+  return disambiguateByMatchInner(dctx, candidates, ignoreWhere, ambiguousBest);
+
+  // TODO: maybe implement the verify procedure in dyno
+  // if (fVerify) {
+  //   // check that 'compareSpecificity' creates a partial order
+  //   PartialOrderChecker checker(candidates, DC);
+
+  //   for (int i = 0; i < candidates.n; ++i) {
+  //     ResolutionCandidate* candidate1         = candidates.v[i];
+  //     bool forGenericInit = candidate1->fn->isInitializer() || candidate1->fn->isCopyInit();
+
+  //     for (int j = i; j < candidates.n; ++j) {
+  //       ResolutionCandidate* candidate2 = candidates.v[j];
+
+  //       int cmp = compareSpecificity(candidate1,
+  //                                    candidate2,
+  //                                    DC,
+  //                                    i,
+  //                                    j,
+  //                                    forGenericInit);
+
+  //       checker.addResult(i, j, cmp);
+  //     }
+  //   }
+
+  //   checker.checkResults();
+  // }
+
+}
 
 
+static MostSpecificCandidates
+disambiguateByMatch(DisambiguationContext& dctx, CandidatesVec& candidates) {
```

Comment:
This is replacing `computeMostSpecificCandidates` / `computeMostSpecificCandidatesWithVecs` but I thought those were more optimized & I wouldn't expect your effort would require changing them. Were you seeing otherwise in testing?

Reply:
I removed the additional procedure and just modified some call sites in these